dace/sdfg/state.py

```diff
@@ -2849,22 +2852,28 @@ def replace_dict(self, repl: Dict[str, str],
     def to_json(self, parent=None):
         return super().to_json(parent)
 
-    def _add_node_internal(self, node, is_continue=False, is_break=False):
-        if is_continue:
-            if is_break:
-                raise ValueError('Cannot set both is_continue and is_break')
-            self.continue_states.add(self.node_id(node))
-        if is_break:
-            if is_continue:
-                raise ValueError('Cannot set both is_continue and is_break')
-            self.break_states.add(self.node_id(node))
-
-    def add_node(self, node, is_start_block=False, is_continue=False, is_break=False, *, is_start_state: bool = None):
-        super().add_node(node, is_start_block, is_start_state=is_start_state)
-        self._add_node_internal(node, is_continue, is_break)
-
-    def add_state(self, label=None, is_start_block=False, is_continue=False, is_break=False, *,
+    def add_state(self, label=None, is_start_block=False, is_break=False, is_continue=False, *,
                   is_start_state: bool = None) -> SDFGState:
         state = super().add_state(label, is_start_block, is_start_state=is_start_state)
-        self._add_node_internal(state, is_continue, is_break)
+        # Cast to the corresponding type if the state is a break or continue state.
+        if is_break and is_continue:
+            raise ValueError('State cannot represent both a break and continue at the same time.')
+        elif is_break:
+            state.__class__ = LoopRegion.BreakState
+        elif is_continue:
+            state.__class__ = LoopRegion.ContinueState
         return state
+
+
+    class BreakState(SDFGState):
```

Comment:
Break is a state transition. However, in a loop region, there is no state transition out of a loop region to signify a break, since that would be an edge from one graph into another. As such, the state transition is implicit, just as it is with the last state of a loop body or the exit state of a nested SDFG. The issue then is that there are three types of implicit end state state transitions: 1) a regular loop body exit state (go back to loop head), 2) a continue exit state (go back to loop head), 3) a break exit state (go to state transitions leading out of loop region). To distinguish which implicit exit state transition to take, there are two special kinds of states for situations 2) and 3). They act as regular states, but when they act as implicit exit states (i.e., no explicit transition leads out of them), their type dictates what type of implicit state transition is required.

Reply:
Is there a concrete exit state (loop condition returns false), or is it implicit?

Reply:
It's implicit - I suppose you're asking because of Python's `for-else` / `while-else` pattern?
If so: that is handled by setting a symbol `__loop_name_did_break` to 1 on transitioning into a `BreakState`, and then inserting state transitions to the Loop's successors based on that symbol value.